SU2_CFD/src/numerics_direct_mean.cpp

```diff
@@ -3993,6 +3993,51 @@ void CSourceGravity::ComputeResidual(su2double *val_residual, CConfig *config) {
   
 }
 
+CSourceBodyForce::CSourceBodyForce(unsigned short val_nDim, unsigned short val_nVar, CConfig *config) : CNumerics(val_nDim, val_nVar, config) {
+
+  /*--- Store the pointer to the constant body force vector. ---*/
+
+  Body_Force_Vector = new su2double[nDim];
+  for (unsigned short iDim = 0; iDim < nDim; iDim++)
+    Body_Force_Vector[iDim] = config->GetBody_Force_Vector()[iDim];
+
+  /*--- Check for compressibility ---*/
+
+  compressible = (config->GetKind_Regime() == COMPRESSIBLE);
+
+}
+
+CSourceBodyForce::~CSourceBodyForce(void) {
+
+  if (Body_Force_Vector != NULL) delete [] Body_Force_Vector;
+
+}
+
+void CSourceBodyForce::ComputeResidual(su2double *val_residual, CConfig *config) {
+
+  unsigned short iDim;
+  su2double Force_Ref = config->GetForce_Ref();
+
+  if (compressible) {
+
+    /*--- Zero the continuity contribution ---*/
+
+    val_residual[0] = 0.0;
+
+    /*--- Momentum contribution ---*/
+
+    for (iDim = 0; iDim < nDim; iDim++)
+      val_residual[iDim+1] = -Volume * U_i[0] * Body_Force_Vector[iDim] / Force_Ref;
+
+    /*--- Energy contribution ---*/
+
+    for (iDim = 0; iDim < nDim; iDim++)
+      val_residual[nDim+1] = -Volume * U_i[iDim+1] * Body_Force_Vector[iDim] / Force_Ref;
```

Comment:
Something is probably not correct here. val_residal[nDim+1] is overwritten in each loop cycle.